### DISCUSSION TITLE
How should environmental settings be configured by default?

### DISCUSSION THREAD

**Discussion by brandonkelly** [craftcms/cms#11865] [Polls]
We are debating how environmental settings like `devMode`, `allowAdminChanges`, and `disallowRobots` should be configured by default, in the [`craftcms/craft`](https://github.com/craftcms/craft/) starter project.

## Option 1: From the config file

The config file checks the `CRAFT_ENVIRONMENT` environment variable, and sets the config setting values accordingly. (This is close to how it [currently works](https://github.com/craftcms/craft/blob/main/config/general.php).)

```php
// config/general.php
$config = GeneralConfig::create()
    ->defaultWeekStartDay(1)
    ->omitScriptNameInUrls();

if (App::env('CRAFT_ENVIRONMENT') === 'dev') {
    $config
        ->devMode()
        ->allowAdminChanges();
}

if (App::env('CRAFT_ENVIRONMENT') !== 'production') {
    $config->disallowRobots();
}

return $config;
```

Pros:
- Pretty clear how the app configuration is expected to change across all environments by looking at `config/general.php`.

Cons:
- Not completely clear how Craft is actually configured for an individual environment, without inspecting its environment variables and cross-referencing with `config/general.php`.
- Config changes require Git commits + deployments.

## Option 2: With environment variables

`config/general.php` is only responsible for defining the default config, leaving all environment-specific changes up to each environment, via environment variables.

```php
// config/general.php
return GeneralConfig::create()
    ->defaultWeekStartDay(1)
    ->omitScriptNameInUrls();
```

```sh
# .env
CRAFT_DEV_MODE=true
CRAFT_ALLOW_ADMIN_CHANGES=true
CRAFT_DISALLOW_ROBOTS=1
```

Pros:
- Easier to understand how Craft is configured for an individual environment, by looking at its environment variables.
- Environmental config changes don’t require a Git commit or deployment.

Cons:
- No way of knowing how each environment is likely to be configured.
- Possibly more error-prone, as it might be easier to forget all of the environment variables that need to be set/adjusted for each environment.

**Comment by aeu** [comment]
Not a vote, but my opinion is that either way is fine, as long as I can put it into version control.

**Comment by michtio** [comment]
Honestly @brandonkelly if you can do:

```
// config/general.php
$config = GeneralConfig::create()
    ->defaultWeekStartDay(1)
    ->omitScriptNameInUrls()
    ->devMode(App::env(DEV_MODE))
    ->allowAdminChanges(App::env(ALLOW_ADMIN_CHANGES)
    ->disallowRobots(App::env(DISALLOW_ROBOTS);

return $config;
```

That would give the best of both worlds in my honest opinion.

**Reply by brandonkelly** [reply]
If you just add `CRAFT_` to the beginning of each of those env variable names, then that is exactly identical to doing:

```php
$config = GeneralConfig::create()
    ->defaultWeekStartDay(1)
    ->omitScriptNameInUrls();

return $config;
```

(In other words, option 2.)

**Reply by mmikkel** [reply]
IMO a big difference between those two approaches is that one of them will keep a record in version control of all the settings being actively configured in the project, and the other one won't.

**Reply by khalwat** [reply]
Yeah I don't think these are even remotely the same, because now someone looking at the `config/general.php` file has no idea what Craft settings are being changed.

They have to know that these "magic" constants exist, and know to look in a separate `.env` file for them.

I think the magic `CRAFT_` settings are not a great idea for reasons expressed below.

**Reply by timkelty** [reply]
> Yeah I don't think these are even remotely the same, because now someone looking at the `config/general.php` file has no idea what Craft settings are being changed.
> 
> They have to know that these "magic" constants exist, and know to look in a separate `.env` file for them.
> 
> I think the magic `CRAFT_` settings are not a great idea for reasons expressed below.

The flip-side is someone one the ops-side doesn't have to know anything about how a particular Craft installation is configured to change a setting.

**Reply by khalwat** [reply]
I think the example given in Option 1 isn't a great one, because looks complicated to people, with a whole bunch of conditionals. It also looks unfamiliar to people who have been using the craftcms/craft starter for some time.

If what you're really trying to decide is whether in the default starter you should have it leveraging the `CRAFT_` magic env vars or not, probably choosing an example that focuses on just _that_ difference would be better.

I know I'd choose option #2 (assuming I didn't know about the other issues with the `CRAFT_` magic env vars) just because it looks simpler, without all of the conditionals, and it looks less verbose 

**Option 1:**

```php
$isDev = App::env('CRAFT_ENVIRONMENT') === 'dev';
$isProd = App::env('CRAFT_ENVIRONMENT') === 'production';

return GeneralConfig::create()
    ->defaultWeekStartDay(1)
    ->omitScriptNameInUrls()
    ->devMode($isDev)
    ->allowAdminChanges($isDev)
    ->disallowRobots(!$isProd)
;
```

**Option 2:**

```php
// config/general.php
return GeneralConfig::create()
    ->defaultWeekStartDay(1)
    ->omitScriptNameInUrls();
```

```env
# .env
CRAFT_DEV_MODE=true
CRAFT_ALLOW_ADMIN_CHANGES=true
CRAFT_DISALLOW_ROBOTS=1
```

**Reply by lindseydiloreto** [reply]
@khalwat This refactor of Option 1 is _much_ cleaner and easier to read. 👍

**Reply by jeromecoupe** [reply]
@khalwat option 1 expressed like this gets my vote: easy to explain and reason about, explicit, flexible, transparent.

**Reply by khalwat** [reply]
Yeah I mean that's my point... that's what they are _already using_ here: https://github.com/craftcms/craft/blob/main/config/general.php

(minus the comments)

If they want a meaningful vote on whether to go with `CRAFT_` magic env vars are not, why not use what they already have, rather than an oddly over-complicated conditional example in option 1?

**Reply by jeromecoupe** [reply]
As a way to fiddle with the fluent flat config file approach, I just updated a production site today to

```php
$isDev = App::env('ENVIRONMENT') === 'dev';
$isProd = App::env('ENVIRONMENT') === 'production';

return GeneralConfig::create()
  ->defaultWeekStartDay(1)
  ->omitScriptNameInUrls()
  ->limitAutoSlugsToAscii(true)
  ->defaultSearchTermOptions([
    'subLeft' => true,
    'subRight' => true,
  ])
  ->devMode($isDev)
  ->allowAdminChanges($isDev)
  ->disallowRobots(!$isProd)
  ->securityKey(App::env('SECURITY_KEY'))
  ->cpTrigger(App::env('CP_TRIGGER') ?: 'admin')
  ->aliases([
    '@web' => App::env('BASE_URL'),
    '@baseUrl' => App::env('BASE_URL'),
    '@basePath' => App::env('BASE_PATH'),
    '@assetsBasePath' => App::env('BASE_PATH').'/uploads',
    '@assetsBaseUrl' => App::env('BASE_URL').'/uploads',
    '@webroot' => App::env('BASE_PATH')
  ])
;
```
Which is pretty explicit and readable (that autocomplete / doc for config variables is a very nice addition). Wrote a similar `db.php` file referencing `.env` vars as well. [Consolidating both in a single config file](https://github.com/craftcms/cms/discussions/11867) as suggested by @michaelrog might be even nicer.

**Comment by mattbloomfield** [comment]
I'm changing my vote - initially I liked the idea of keeping the current config - but the more I think about it the more I think this is the only file that defines things for multiple environments. The rest of the Craft ecosystem only defines the environment you are working in currently. I think for the sake of consistency and understanding it's actually probably better to handle configuration at the .env level.

**Comment by jeromecoupe** [comment]
Currently, config files with simple logic and lean `.env` allow for very flexible setups.
I like to keep my`.env` file lean and only use those vars for:

1. Environment type (dev, staging, production)
1. Sensitive data (credentials, secrets, etc.)
2. Variables changing from environment to environment (paths, urls, etc.)

If we are talking about a default for new projects, my vote would go to a few `.env` vars to setup, the rest being handled by a config file referencing those with sensible defaults for cross environment variables.

When teaching to newbies at school, I feel a solid config file with simple logic + a minimal amount of `.env` vars to setup makes the most sense to them: the config file does the heavy lifting, the `.env` file is only for what needs to change from environment to environment.

**Reply by brandonkelly** [reply]
> 3. Variables changing from environment to environment (paths, urls, etc.)

That’s exactly the type of config settingns in question here. Per the second example, we would continue recommending using `config/general.php` to define settings that don’t need to change per-environment.

**Reply by jeromecoupe** [reply]
I am mainly with @khalwat here. This is a front-end dev perspective, so maybe take it with a grain of salt

- I like config files to explicitly reference all required variables so I can have a clear overview. To me config files are the source of truth. `.env` variables should not override anything set in config files. 
- a limited amount of `.env` variables (explicitly referenced in config files) allow for multi environment configs while keeping config files flat with only very simple logic (flags, string concat, etc.).
- IMHO `CRAFT_` prefixed .env vars being able to override anything is a devops super power, including for DB credentials. Newbies and devs should not have to use those.

To sum things up for default craft installs:

- everything is explicitly set in flat config files, including all references to `.env` vars
- `.env` vars are only used for credentials and environment specific vars and cannot override anything

**Comment by khalwat** [comment]
Please don't go all environment variables. I discuss why in detail here: https://nystudio107.com/blog/fluent-multi-environment-config-for-craft-cms-4#config-settings-via-environment-variable

Environment variables were originally intended for local development only, but everyone started using them in production anyway. Environment variables are intended to be for things that vary from environment to environment.

Putting _every_ setting into environment variables seems like shoehorning settings into a place where they were never intended to be.

More importantly, option 2 splits the config between two separate files, so you have to hunt through both to figure out where a setting is being set. One file invisibly overrides the other, which is not great from a transparency POV.

I can easily envision a situation where someone is explicitly setting `->allowAdminChanges(false)` in their `config/general.php` file and they are completely confused as to why it's not working. They can see it there plain as day. But it's invisibly being overridden by a change someone made in the `.env` file with `CRAFT_ALLOW_ADMIN_CHANGES=true`

There would be no indication that the setting is being overridden via the `.env` file, nothing linking the settings in the `config/general.php` file that they are used to looking at to the `.env` file... so they'd just have to know that they need to look in both places, which doesn't seem very transparent to me

By having the default project working this way, you're setting people up to have issues, imo.

Also with any change, you have to consider the benefit of the change vs. the cost of it. I'm not seeing the benefit outweighing the cost here.

I strongly agree with @michtio's comment. We can already do this:

```php
// config/general.php
$config = GeneralConfig::create()
    ->defaultWeekStartDay(1)
    ->omitScriptNameInUrls()
    ->devMode(App::env(DEV_MODE))
    ->allowAdminChanges(App::env(ALLOW_ADMIN_CHANGES)
    ->disallowRobots(App::env(DISALLOW_ROBOTS);

return $config;
```

Not seeing the benefit of either option 1 and especially not option 2

The best solution here is changing nothing, imo

**Reply by michtio** [reply]
Thanks for the extended answer.

**Reply by lindseydiloreto** [reply]
I agree, this is by far the best approach. It's the best of all worlds, easiest to read & easiest to manage.

**Reply by brandonkelly** [reply]
> Please don't go all environment variables.

That’s not remotely suggested. We will definitely continue defining things that don’t need to change per-environment in `config/general.php`.

> Environment variables were originally intended for local development only, but everyone started using them in production anyway. 

Not sure where you’re getting that from? They’ve always been needed for storing API keys, etc., at the very least.

> More importantly, option 2 splits the config between two separate files, so you have to hunt through both to figure out where a setting is being set.

Fair, and mentioned in the Cons for option 2.

**Reply by khalwat** [reply]
> That’s not remotely suggested. .

Well they are already there as magic hidden overrides to config settings, which I think is problematic for reasons discussed below (see my replies to @lindseydiloreto)

And option 2 encourages that behavior. I'd be happy to see the magic `CRAFT_` overrides removed, or at the very least, some way to disable them entirely. Maybe a `CRAFT_ENV_MAGIC` constant or something.

> Not sure where you’re getting that from? They’ve always been needed for storing API keys, etc., at the very least

As `.env` files read in via phpdotenv I mean, not as injected environment variables (see discussion below with Jason)

**Comment by jasonmccallister** [comment]
> Environment variables were originally intended for local development only, but everyone started using them in production anyway. Environment variables are intended to be for things that vary from environment to environment.

I’m not sure I agree with that statement.. following things like 12factor apps and deploying to containers… everything is defined at an environment level or secret (depending upon the platform). 

If someone messed up a production config and disabled robots changing an environment variable to override that makes a lot of sense to me.. especially if the company is large and broken into multiple teams… an ops guy doesn’t want to touch code and redeploy to potentially break production, but in most platforms it’s easy to change an environment to toggle something on and then off again to change the behavior of the app.

Ultimately it depends on the context, single person shop vs agency vs large team/enterprise.

It actually feels like command line flag parsing in CLI applications… the existence of the flag overrides the environment variable.

**Comment by khalwat** [comment]
> I’m not sure I agree with that statement.. following things like 12factor apps and deploying to containers… everything is defined at an environment level or secret (depending upon the platform).

The phpdotenv developer had it in all caps in his README.md for years not to use it in production, which is how I suspect most people are using `.env` today with Craft, rather than via environment injection.

Obviously, it's different in terms of "using it on production" if we're talking about injecting the variables into the environment via Docker container or via Nginx, or what have you.

As to "everything" being defined at an environment level, I'm just not sure I love the idea of all 158 config settings for Craft CMS (not to mention the DB settings) being defined via `.env` config (yes, I know this is not what option 2 is advocating for, just addressing your comment), for the reasons mentioned in my previous post.

It's also a whole lot more error-prone to do it that way, you have to convert the setting from the Craft setting into SCREAMING_SNAKE_CASE and you get no autocomplete, documentation, or type checking when doing so (which was introduced by the fluent config)

**Reply by jasonmccallister** [reply]
Oh, yeah your post wasn't clear.. I did not mean using an actual .env file.. I mean the platforms environment variable system (ECS, Kubernetes, Cloud Run, and etc).

No one should be using a .env in production.. I also agree all settings should not be defined in the .env but have the capability to override using an environment variable if that makes sense?

The defaults for a config should be sensible.. 

Does that make more sense?

**Comment by khalwat** [comment]
> No one should be using a .env in production..

Agreed; but I'd wager that a majority of Craft CMS sites are in fact using `.env` via phpdotenv in production today. I even wrote this to try to help ameliorate that: https://github.com/nystudio107/dotenvy

> I also agree all settings should not be defined in the .env but have the capability to override using an environment variable if that makes sense?

It does, but I am not a fan for the transparency scenario mentioned above (re-pasted here):

I can easily envision a situation where someone is explicitly setting `->allowAdminChanges(false)` in their `config/general.php` file and they are completely confused as to why it's not working. They can see it there plain as day. But it's invisibly being overridden by a change someone made in the `.env` file with `CRAFT_ALLOW_ADMIN_CHANGES=true`

There would be no indication that the setting is being overridden via the `.env` file, nothing linking the settings in the `config/general.php` file that they are used to looking at to the `.env` file... so they'd just have to know that they need to look in both places, which doesn't seem very transparent to me

By having the default project working this way, you're setting people up to have issues, imo.

**Comment by michaelrog** [comment]
> We are debating how environmental settings like `devMode`, `allowAdminChanges`, and `disallowRobots` should be configured by default, in the [craftcms/craft](https://github.com/craftcms/craft/) starter project.

Short answer — Either method is probably fine, _for the specific configs you listed_.

In my projects, I define these three configs in the PHP config, based on a single env var: 

```
$inDev = App::env('CRAFT_ENVIRONMENT') === 'dev';

return GeneralConfig::create()
    ->devMode($inDev)
    ->allowAdminChanges($inDev)
    ->disallowRobots($inDev);
```

I specifically prefer to put these in the PHP config, because:
- Then I have a "single source of truth" rather than splitting config across several files. — I can grok the shape of the config, and what env vars I expect to find, at a glance.
- These three configs toggle in tandem, so I can derive the value from a single env var.

(Notably, the `CRAFT_ENVIRONMENT` constant is a carryover from previous versions of Craft, and it's no longer set by default in the Craft 4 version of `craftcms/craft`. I've opted to keep defining it in my projects anyway.)

---

Long answer — Even though it'd probably be fine to treat some configs as exceptional overrides, it's still preferable not to: The `craftcms/craft` starter project establishes conventions that a lot of devs copy, especially novices and folks new to the ecosystem. **In general, I think most devs will be well-served by a strong preference for PHP configs — including environment-specific values that are used across multiple envs.**

You've hit the nail on the head here:
> config/general.php is ... responsible for defining the default config...

...but I draw slightly broader boundaries around what "default" means. **To me, a _default_ config is anything that I _expect_ to be present across multiple environments.**

For example, it may be true that the _value_ of my default site URL changes from env to env... but I still consider this part of my default setup because _every env has/needs this value_.

**The benefits of version control, a single mental source-of-truth, and fluent IDE auto-completion, make leaning on PHP configs my preference for the vast majority of configs.**

(And, I don't want to clutter up my `.env` file. The _env_ isn't here to configure _Craft_ so much as to configure the _process_. The schema of _Craft's_ config is best represented in _Craft's_ config file. `.env` is more of an external value loader; We shouldn't overload its responsibility, if we can help it.)

All that said, I _can_ envision use cases where reading config directly from the `.env` _could_ be useful:
 - Values that are temporary/transient, that are specifically intended to _override_ the config file without changing it
 - Values that represent real exceptions, i.e. are only present in special cases, and I wouldn't expect to see them used across environments
 - e.g. Tinkering with configs in-place, without triggering a deployment
 - e.g. For hosts like Servd (or Craft Cloud?), allowing an external control panel to toggle something like `isSystemLive` without changing user-defined files
 - e.g. ephemeral/serverless setups, where we want to configure different instances of Craft programmatically by manipulating process vars

So, it's definitely worth raising awareness that this method exists, but only in special cases.

Once a config value — whether static or environment-specific — is expected or shared in multiple envs, I typically consider it to be worthy of committing to the PHP config file. (And I think this preference is worthy of elevating as a best-practice, _especially_ in `craftcms/craft`.)

Admittedly, my preference is biased by living in agency-world for most of my career, where I highly prize being able to (a) infer the likely config of a site at a glance, from the code, and (b) easily onboard new devs, especially novices, to lots of different projects. I find it's more practical to optimize for human understanding. From that standpoint, the cons of Option 2 outweigh the cons of Option 1, if we're talking about establishing a default convention that _most_ sites are going to use out of the box.

**Reply by khalwat** [reply]
@michaelrog wrote:

> * Values that are temporary/transient, that are specifically intended to _override_ the config file without changing it
> * Values that are only present in special cases, that need to override the default
> * e.g. Tinkering with configs in-place, without triggering a deployment
> * e.g. For hosts like Servd (or Craft Cloud?), allowing an external control panel to toggle something like `isSystemLive` without changing user-defined files
> * e.g. ephemeral/serverless setups, where we want to configure different instances of Craft programmatically by manipulating process vars
> 
> Those cases should be _exceptional_, though, and typically represent more advanced usage.

Agree with all of these scenarios (and that they are _exceptional_), but you can still get all of that with the current config by doing this:

```php
// config/general.php
$config = GeneralConfig::create()
    ->defaultWeekStartDay(1)
    ->omitScriptNameInUrls()
    ->devMode(App::env(DEV_MODE))
    ->allowAdminChanges(App::env(ALLOW_ADMIN_CHANGES)
    ->disallowRobots(App::env(DISALLOW_ROBOTS);

return $config;
```

You can then still tinker with the environment variable settings to your heart's content, and it'll all work.

The key difference being **it is not an invisible override** of the config settings. It's a transparent one.

I'd honestly be in favor of eliminating the `CRAFT_` prefix that magically overrides explicit settings in the `config/general.php` altogether.

If you need to do that, put the setting into an environment variable that the `config/general.php` file explicitly reads from. Hidden magic is no bueno.

**Reply by michaelrog** [reply]
Yeah, the key is whether you expect to be setting a config ahead of time (so you write it into your config file and deploy it), vs tinkering at runtime.

I like being able to inject runtime overrides, and I think the `CRAFT_` prefixed envs are a clever way to accomplish this when needed. I just wouldn't want to enshrine it as the go-to option for _most_ config.

**Reply by khalwat** [reply]
I hear ya. I think that's an edge case of an edge case tho, most likely. And I think it's better to be intentional about it when you do, rather than magic overrides.

**Comment by lindseydiloreto** [comment]
Are you really going to introduce a "magic" prefixed `CRAFT_` environmental variable for ~150 different configuration settings?

That seems pretty extreme, especially considering that the list could grow or change over time. It made much more sense to have magic environmental variables for the **database** configuration. There are only about 8-10 database settings (which is probably all you'll ever need), and DB settings are _highly_ environment-dependent.

I just don't think you can say "this helped with `db.php`, we should apply the same logic to `general.php`."

**Reply by khalwat** [reply]
> Are you really going to introduce a "magic" prefixed CRAFT_ environmental variable for ~150 different configuration settings?

It's already there. https://github.com/craftcms/cms/pull/10573

**Reply by lindseydiloreto** [reply]
Yikes, I totally missed that. I don't like it (with the exception of DB values), so I guess I just won't use it.

**Reply by khalwat** [reply]
Yeah I'm not a fan either. Here's what it does:

https://github.com/craftcms/cms/blob/ac616d429ee369b83478b3e8441ed5d6693f9036/src/helpers/App.php#L115

It iterates over every property that `General.php` has, and then via `App::env()` checks to see if a properly named environment variable exists, and if so, it overrides the config setting with whatever is in that variable.

So even if you don't use it at all it's checking every property every time.

Mostly I'm not a fan of the lack of transparency this introduces (scenarios mentioned in above posts)

**Reply by timkelty** [reply]
> It iterates over every property that General.php has, and then via App::env() checks to see if a properly named environment variable exists, and if so, it overrides the config setting with whatever is in that variable.

When this was implemented, I did some fairly extensive performance profiling and the overhead was barely measurable. Performance should not be a concern here.

**Reply by khalwat** [reply]
Yes, I saw the performance profiling, and noted that while it is 20x slower than without it, we're talking nanoseconds. So it's not an issue. I wasn't mentioning performance, but rather that you have no choice, it does this for all of the settings regardless of whether you use it at all, or want to use it.

**Comment by mmikkel** [comment]
To throw my hat in the ring: 

I think the config override feature should be considered just that: a way to *override* config, without having to commit or deploy code - i.e., being able to toggle something like `devMode` in production in a pinch – not the preferred way to configure (parts of) Craft.

Even if we're currently mostly talking about config settings that are environmental in nature (like `allowAdminChanges`, `devMode` and even the database config), IMO it's likely a bad idea to start encouraging devs to configure Craft using settings that are _only_ defined using env vars. It's really going to suck a few years down the line, when that server blows up and nobody can even remember what settings they were _using_, let alone their values. Even if the actual values can't be version controlled, at the very least having a record of the settings at play by including them in `general.php` and using `App::env()` etc. to pull in the values, is going to be way more safe in the long term, and should be encouraged as best practice.

**Reply by lindseydiloreto** [reply]
I fully agree. I'd like to be able to look in the `general.php` file and see which values are coming from the environment itself.

If someone only sets a `CRAFT_` override without any mention of it in the config file, that is "magical" behavior which could be very problematic down the road. A future developer on the project may lose many hours trying to debug a mysterious issue.

**Reply by mmikkel** [reply]
Right, or – migrating the project to a new server, and missing a crucial config setting in the new production environment because nobody knew it was even set in the old one.

**Reply by khalwat** [reply]
I agree, though I'll go even further to say that I think `CRAFT_` overrides being able to override any config setting is problematic, in that it can cause some serious confusion.

So if I have in my config file: 
```
    ->devMode(false)
```
 it's possible for `devMode` to be set to `true` if someone puts: 
```
CRAFT_DEV_MODE=true
``` 
in the `.env` file. I can see some people being very confused as to why their changes to the `config/general.php` file are not working, and it being because of these magic overrides.

Being explicit and intentional in the `config/general.php` file, by declaring that the setting comes from an environment variable is going to result in fewer headaches, IMO.

And you can still override the `.env` setting in a pinch if you need to. The difference is clarity in terms of what is going on.

**Reply by lindseydiloreto** [reply]
Even just reading this thread is giving me gray hair. 😬

**Comment by timkelty** [comment]
There are clearly a lot of strong opinions in both directions here, which is understandable, as I think it is really personal preference.

I think it really boils down to:
- people that are more concerned with the ops and deployment of a Craft app may prefer to use env overrides,
- those more concerned with the app itself may prefer to have more things in `config/general.php`

Personally, I love the magic env (admittedly biased) – as I often support Craft apps I didn't build, and I want to be able to set something without combing through a `config/general.php` to see how something can be configured. Furthermore, the ability to do this remotely without a deployment and 💯 know it will work regardless of the app is a huge win for me.

**Reply by mmikkel** [reply]
Config overrides is potentially a massively useful feature and I can think of tons of use cases for them. There are some big caveats involved though and they feel like a power user kind of thing to me – I just don't know if they make the most sense for a default config in the starter project; something that is probably used as the boilerplate for most Craft projects by developers at all different kinds of levels.

This feature is also available regardless of what the project's config file actually looks like. If the concern is that the ability to configure Craft with `CRAFT_`-prefixed env vars will go unnoticed by devs (or ops) who will find that approach preferable, why not include some comments in the starter project's `general.php` file, pointing to the feature in the docs?

**Reply by khalwat** [reply]
I'm equally comfortable doing devops and doing Craft development. But I think most Craft developers are better served by clear, transparent ways of configuring Craft.

Screaming snake case magic env vars I think have too many potential gotchas, and too many downsides compared to relatively niche devops benefits.

The primary benefit I see from them is being able to override a Config setting that is not in the general.php file pulling from an environment variable.

Most of the time, the env vars you might want to change on the fly are already environment-specific, and thus would already be coming from and environment variable via general.php

The relatively rare times that this isn't the case I don't think outweighs all of the negatives that come along with using them, for the majority of Craft developers

**Reply by khalwat** [reply]
Ooops, I take it back -- I forgot that I asked to have this changed/fixed in https://github.com/craftcms/craft/commit/e84b53ec35509fa78ae7c03eb9c838f9a01b1abe

**Comment by lindseydiloreto** [comment]
I believe this statement by @timkelty (emphasis mine)...

> **I often support Craft apps I didn't build**, and I want to be able to set something without combing through a `config/general.php` to see how something can be configured. Furthermore, **the ability to do this remotely without a deployment** and 💯 know it will work regardless of the app is a huge win for me.

... and this statement by @mmikkel (emphasis mine)...

> Config overrides is potentially a massively useful feature and I can think of tons of use cases for them. There are some big caveats involved though and they **feel like a power user kind of thing to me**.

... are in _complete and total agreement_. They are both talking about the `CRAFT_` prefix as an excellent tool for use-cases **beyond** ordinary circumstances. It is a powerful weapon for people who need to ninja their way into a staging/production server.

But for your typical day-to-day Craft usage, it seems excessive. So let's go back to the original question (emphasis mine)...

> We are debating how environmental settings... should be configured **by default, in the `craftcms/craft` starter project.**

### My conclusion...

Designed as a tool for power users in extraordinary circumstances, the `CRAFT_` prefix should probably **not** be an integral part of the general configuration for each default Craft installation. It should be reserved for special usage, when it is necessary to alter Craft configurations in a new and powerful way.

With the power of `CRAFT_` env variables, there is a very real risk of [sending future developers down a rabbit hole chasing mysterious bugs](https://github.com/craftcms/cms/discussions/11865#discussioncomment-3531167). It would be prudent to keep that tool out of the default configuration.

---

### EXCEPTION - Database Config

For all that I said above regarding the **general** config settings, I don't believe the same logic applies to **database** config settings. The DB configuration settings are a different scenario with very different considerations...

- Using env variables allows you to eliminate the `config/db.php` file (which isn't true for `config/general.php`).
- The list of `CRAFT_DB_` variables is limited to a very small number, and unlikely to ever change.
- Database values are **explicitly different per each environment**.

Despite what I said about the general config, it **does** make sense to use `CRAFT_DB_` for the database config settings.

**Reply by jeromecoupe** [reply]
Agree with pretty much everything above regarding `CRAFT_` (devops super power, stay clear of them in default configs). I wouldn't make an exception for DB credentials in default installs, though. Exceptions make things more difficult to explain to new devs. I would keep things as simple and straightforward as possible:

- everything is explicitly set in flat config files, including all references to `.env` vars
- `.env` vars are only used for credentials and environment specific vars and cannot override anything

**Comment by rostockahoi** [comment]
Personally I don't like the magic `CRAFT_` env vars. I would strongly prefer option 1.

As a side note, here is a solution that would weaken the first con point of option 1:

> Not completely clear how Craft is actually configured for an individual environment, without inspecting its environment variables and cross-referencing with config/general.php.

Laravel recently introduced a new `about` command. The command resolves the actual configuration so it takes away the burden of manually doing it. The output looks like this:

![image](https://user-images.githubusercontent.com/16897125/188124309-23693431-447e-4dad-8655-72dee8eeb899.png)

**Reply by khalwat** [reply]
I liked it! Two points here:

1 - A Utility & CLI command that would show you all of the config settings _that are different from the default_, their values (redacted as needed), and critically _where they are being set from_ (which config file or which `CRAFT_` magic env var) would be a useful addition regardless of what happens with the default config

2 - I've been hoping for something similar to Laravel's `artisan config:cache` that resolves the config for some time: https://github.com/craftcms/cms/discussions/8418

**Comment by dominikkrulak** [comment]
Reading all these comments here with good points, pros and cons and thinking about final decision that @brandonkelly and his team behind Craft CMS must make must be frustrating. I think I get it that @brandonkelly wants Craft CMS to be great UX for developers, content creators etc. and best CMS in the world but

I'd always look at it from performance point of view even if a difference is bearable.
**The less overwrite the better.**

... plus more settings come in a future.

No matter how luxurious for developer spinning up a new Craft CMS's install would be, we build stuff for end user (visitor of a website or an app). Craft CMS should be fast and ecological in processing a request.

Doing one more round just to check if there are some more settings and those settings precedes other settings is kind of `sudo` power with pros and cons mentioned above.

Personally I wouldn't go for overriding settings with `CRAFT_` prefix. I like clean approach to a solution and I don't like overriding.

Although I would like `CRAFT_` prefix only if it wouldn't override anything. Like for example current `general/db.php` file. Now you have option to choose to go with `CRAFT_` prefix which eliminates the need of `general/db.php` file or to go with `general/db.php` file. Craft still needs to check for both. I think it would be better if Craft would look for connection settings only in one place like `.env` file.

Based on what the option 2. would do under the hood I choose performance over developer's comfort and therefore 
I lean towards option 1. `general/config.php` and `.env` file.

**Comment by brandonkelly** [comment]
Thanks to everyone for chiming in!

At the end of the day, [this comment](https://github.com/craftcms/cms/discussions/11865#discussioncomment-3531354) (and others with a similar sentiment) won me over, for ruling out option 2:

> IMO a big difference between those two approaches is that one of them will keep a record in version control of all the settings being actively configured in the project, and the other one won't.

We decided to go with @michtio’s [hybrid approach](https://github.com/craftcms/cms/discussions/11865#discussioncomment-3529831) (list out each of the settings in use in `config/general.php`, and have them pull in environment variables when needed).

From [config/general.php](https://github.com/craftcms/craft/blob/main/config/general.php):

```php
return GeneralConfig::create()
    // Set the default week start day for date pickers (0 = Sunday, 1 = Monday, etc.)
    ->defaultWeekStartDay(1)
    // Prevent generated URLs from including "index.php"
    ->omitScriptNameInUrls()
    // Enable Dev Mode (see https://craftcms.com/guides/what-dev-mode-does)
    ->devMode(App::env('DEV_MODE') ?? false)
    // Allow administrative changes
    ->allowAdminChanges(App::env('ALLOW_ADMIN_CHANGES') ?? false)
    // Disallow robots
    ->disallowRobots(App::env('DISALLOW_ROBOTS') ?? false)
;
```

We‘ve also replaced the one `.env.example` file with three example files, which demonstrate how we’d expect the `DEV_MODE`, `ALLOW_ADMIN_CHANGES`, and `DISALLOW_ROBOTS` environment variables to be set depending on the environment type:

File | `DEV_MODE` | `ALLOW_ADMIN_CHANGES` | `DISALLOW_ROBOTS`
---- | :--------- | :-------------------- | :----------------
[.env.example.dev](https://github.com/craftcms/craft/blob/main/.env.example.dev) | `true` | `true` | `true`
[.env.example.staging](https://github.com/craftcms/craft/blob/main/.env.example.staging) | `false` | `false` | `true`
[.env.example.production](https://github.com/craftcms/craft/blob/main/.env.example.production) | `false` | `false` | `false`

`craftcms/craft` 2.0.2 has been tagged with these changes. (craftcms/craft#87)
